Applications/Utils/ModelPreparation/PartitionMesh/MeshPartitioning.cpp

```diff
@@ -0,0 +1,493 @@
+                                      "Number of integers to define elements; Number of integers to define ghost elements; "
+                                      "Reserved number.";
+        os_subd_head << mesh_info << std::endl;
+        os_subd_head << npartitions << std::endl;
+
+        fname = file_name_base + "_partitioned_elems_" + npartitions_str + ".msh";
+        os_subd.open(fname.c_str(), std::ios::out|std::ios::trunc );
+
+        fname = file_name_base + "_partitioned_nodes_" + npartitions_str + ".msh";
+        os_subd_node.open(fname.c_str(), std::ios::out|std::ios::trunc );
+
+        std::setw(14);
+        os_subd.precision(14);
+        os_subd.setf(std::ios::scientific);
+    }
+
+    PetscInt global_node_id_offset = 0;
+    std::vector<bool> nodes_part_reserved;
+    nodes_part_reserved.reserve(nnodes);
+    std::vector<PetscInt> node_global_ids;
+    node_global_ids.reserve(nnodes);
+    for (std::size_t i=0; i<nnodes; i++)
+    {
+        nodes_part_reserved.push_back(false);
+        node_global_ids.push_back(0);
+    }
+    std::vector<bool> nodes_reseved(nnodes);
+    std::vector<bool> elems_reseved(_elements.size());
+
+    std::vector<MeshLib::Node*> partition_nodes;  // non-ghost nodes of a partition
+    std::vector<std::size_t> partition_start_node_id(npartitions);
+    std::vector<std::size_t> partition_end_node_id(npartitions);
+    std::vector<std::size_t> end_non_ghost_node_id(npartitions);
+
+    const PetscInt num_headers = 14;
+    PetscInt head[num_headers];
+    // node rank offset
+    head[10] = 0;
+    // element rank offset
+    head[11] = 0;
+    // ghost element rank offset
+    head[12] = 0;
+    // Reserved
+    head[13] = 0;
+
+    for (std::size_t ipart=0; ipart<npartitions; ipart++)
+    {
+        INFO("Processing partition: %d", ipart);
+
+        partition_start_node_id[ipart] = partition_nodes.size();
+
+        std::fill(nodes_reseved.begin(), nodes_reseved.end(), false);
+
+        // Find non-ghost nodes in this partition
+        for (std::size_t i=0; i< nnodes; i++)
+        {
+            if ( (nodes_partition_ids[i] == ipart) && (!nodes_part_reserved[i]) )
+            {
+                partition_nodes.push_back(_nodes[i]);
+                nodes_part_reserved[i] = true;
+                nodes_reseved[i] = true;
+            }
+        }
+        end_non_ghost_node_id[ipart] = partition_nodes.size();
+
+        // Find elements in this partition
+        std::vector<const Element*> partition_regular_elements;
+        std::vector<const Element*> partition_ghost_elements;
+        std::fill(elems_reseved.begin(), elems_reseved.end(), false);
+        // For ghost element
+        std::vector< std::vector<unsigned> >
+        elems_non_ghost_nodes_local_ids(_elements.size());
+
+        const PetscInt num_non_ghost_nodes =  end_non_ghost_node_id[ipart]
+                                           - partition_start_node_id[ipart];
```

Comment:
Yes, it is different. I will rewritten this loop with mode clear definitions of partition variables, which  will be local with each partition.